### DISCUSSION TITLE
Unable to connect to google cloud sql - postgres db

### DISCUSSION THREAD

**Discussion by m1m6** [prisma/prisma#4170] [Q&A] [answered]
The following message i'm getting when trying to connect to gcloud postgres db using prisma connector:
```
Prisma Schema loaded from prisma/schema.prisma
Error: P1001: Can't reach database server at `/cloudsql/prefab-wave:europe-west3:db`:`5432`

Please make sure your database server is running at `/cloudsql/prefab-wave:europe-west3:db`:`5432`.
```

here's my connection config:
```

datasource db {
  provider = "postgresql"
  url      = "postgres://USER:PASSWORD@localhost/testdb?host=/cloudsql/prefab-wave:europe-west3:db"
}

```

**Comment by ryands17** [comment]
Hey @m1m6 👋 

I think the colons in the URL might be causing the issue. Could you check if the URL string is correct? If it is, I would suggest passing a URL encoded version of the part after the host and check again?

For e.g.

```prisma
datasource db {
  provider = "postgresql"
  url      = "postgres://USER:PASSWORD@localhost/testdb?host=/cloudsql/prefab-wave:europe-west3:db"
}
```

Could you pass a URL encoded version of this `/cloudsql/prefab-wave:europe-west3:db` part and check?

**Comment by m1m6** [comment]
Hi @ryands17 
I've tried with the following encoded URI

`"%2Fcloudsql%2Fprefab-wave%3Aeurope-west3%3Adb"
`i got the same error

**Comment by m1m6** [comment, marked as answer]
well just tried to append the db server name:
```
datasource db {
  provider = "postgresql"
  url      = "postgres://USER:PASSWORD@localhost/testdb?host=DB_SERVER"
}
```

and it works fine 🤦

**Reply by janpio** [reply]
What does that look like in practice?

**Reply by m1m6** [reply]
I have a google cloud Postgres DB, and the DB_SERVER is the public IP address of the database instance 

```
datasource db {
  provider = "postgresql"
  url      = "postgres://USER:PASSWORD@localhost/testdb?host=35.000.000.00"
}
```

**Reply by m1m6** [reply]
btw I had to whitelist the VM server IP against the database instance to make it work.

**Reply by ppichier** [reply]
Hey @m1m6 , thanks for your thread. It helps me a lot. But now i ve got the following error:
`User `postgres` was denied access on the database `postgres.public`

Do you have any idea ?

**Reply by ghost** [reply]
me too i encounter this error, User postgreswas denied access on the databasepostgres.public`

**Reply by LevPewPew** [reply]
after so much time with answers/documentation telling me to have my URL like in the original post, this answer finally fixed my problem.

i was attempting to connect to a PostgreSQL database hosted GCP with a NextJS prisma app deployed on Vercel.

**Reply by sumitra19jha** [reply]
Worked for me postgres://USER:PASSWORD@localhost/testdb?host=35.000.000.00

**Comment by draconteus** [comment]
This format worked for me, if you don't want to hard code the IP address:
```
DB_USER=dbuser
DB_PASS=dbpass
DB_HOST=localhost
DB_PORT=5432
CLOUD_SQL_CONNECTION_NAME=/cloudsql/gcp-project-id:europe-west3:db-instance-name
DATABASE_URL=postgres://${DB_USER}:${DB_PASS}@${DB_HOST}:${DB_PORT}/${DB_BASE}?host=${CLOUD_SQL_CONNECTION_NAME}
```
It was used in Cloud Run application with configured Cloud SQL public connection (with Cloud SQL proxy).

**Reply by DavidLangva** [reply]
I was experiencing the same issue when trying to run `npx prisma db pull`

I tried to use the above format with the `host` parameter, but it wouldn't work.

However, cutting it back a bit worked and the following format did the trick:
```
DB_USER=dbuser
DB_PASS=dbpass
DB_HOST=34.000.000.0
DB_PORT=5432
DATABASE_URL=postgres://${DB_USER}:${DB_PASS}@${DB_HOST}:${DB_PORT}
```

**Reply by bowis** [reply]
How did you guys make Google Cloud Run access the .env file?

**Reply by anshu-mages** [reply]
What would be the DB HOST, would it be the private ip ?

**Reply by sahin52** [reply]
What is DB_BASE? @draconteus

**Comment by FedeBev** [comment]
I've been struggling with the same issue for hours, my problems was that `sqladmin.googleapis.com` service was not enabled.

I just needed to add the following terrafom resouce
```tf
resource "google_project_service" "sqladmin" {
  service = "sqladmin.googleapis.com"

  timeouts {
    create = "5m"
    update = "5m"
  }

  disable_dependent_services = true
  disable_on_destroy = true
}
```

**Reply by thejoecode** [reply]
Thank you. I just searched in google cloud console for `sqladmin.googleapis.com` and clicked the enable button.

**Comment by antoinebourayne** [comment]
Hi i'm still struggling to get prisma connected to my db with the same error : 

`Error: P1001: Can't reach database server at 34.XXX.XXX.XXX:5432`

My prisma config : 

```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id    Int    @id
  login String @unique
}
```

```
import { PrismaClient } from '@prisma/client'

async function main() {
  const client = new PrismaClient()

  const results = await client.user.findMany()
  console.log(results)
}

void main()

```

To reproduce, just create a new db on cloud SQL and run a proxy to connect locally to the instance : 

```
docker pull gcr.io/cloudsql-docker/gce-proxy:1.11
docker run -d -v [PATH_TO_KEY_FILE]:/config \
    -p 127.0.0.1:5432:5432 -p 127.0.0.1:3306:3306 \
    gcr.io/cloudsql-docker/gce-proxy:1.11 /cloud_sql_proxy \
    -instances=[INSTANCE_CONNECTION_NAME]=tcp:0.0.0.0:5432 -credential_file=/config
```

```
DATABASE_URL=postgres://user:password@localhost/db?host=project:region:instance
npx prisma migrate dev
```

Note that doing : 

`psql -h localhost -p 5432 -U user -d db
`
works fine, i believe this has something to do with prisma connection to the Unix socket

**Reply by roland-kiraly-dentsu** [reply]
> Hi i'm still struggling to get prisma connected to my db with the same error :
> 
> `Error: P1001: Can't reach database server at 34.XXX.XXX.XXX:5432`
> 
> My prisma config :
> 
> ```
> generator client {
>   provider = "prisma-client-js"
> }
> 
> datasource db {
>   provider = "postgresql"
>   url      = env("DATABASE_URL")
> }
> 
> model User {
>   id    Int    @id
>   login String @unique
> }
> ```
> 
> ```
> import { PrismaClient } from '@prisma/client'
> 
> async function main() {
>   const client = new PrismaClient()
> 
>   const results = await client.user.findMany()
>   console.log(results)
> }
> 
> void main()
> ```
> 
> To reproduce, just create a new db on cloud SQL and run a proxy to connect locally to the instance :
> 
> ```
> docker pull gcr.io/cloudsql-docker/gce-proxy:1.11
> docker run -d -v [PATH_TO_KEY_FILE]:/config \
>     -p 127.0.0.1:5432:5432 -p 127.0.0.1:3306:3306 \
>     gcr.io/cloudsql-docker/gce-proxy:1.11 /cloud_sql_proxy \
>     -instances=[INSTANCE_CONNECTION_NAME]=tcp:0.0.0.0:5432 -credential_file=/config
> ```
> 
> ```
> DATABASE_URL=postgres://user:password@localhost/db?host=project:region:instance
> npx prisma migrate dev
> ```
> 
> Note that doing :
> 
> `psql -h localhost -p 5432 -U user -d db ` works fine, i believe this has something to do with prisma connection to the Unix socket

I have the same issue. It worked with postgres://user:password@localhost/db?host=project:region:instance in the past, but I cant' get it to work now. Did you found a solution?

**Comment by cheapestorderconcept** [comment]
I am having this particular error too can't access database 34.XXX Have tried all the solutions here nothing work for me. Please I need help. I am working with app engine, prisma and cloud sql(postgress)

**Reply by digitaldeus** [reply]
+1 Just ran into this issue followed all the instructions from above but I cannot run `migrate deploy` for GCP postgres in production.  Am able to connect fine with the user on both command line and TablePlus gui

> Error: P1010: User `postgres` was denied access on the database `postgres.public`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
